site-src/geps/gep-735.md

```diff
@@ -0,0 +1,170 @@
+# GEP-735: TCP and UDP addresses matching
+
+* Issue: [#735](https://github.com/kubernetes-sigs/gateway-api/issues/735)
+* Status: Provisional
+
+## TLDR
+
+Spec for matching source and destination addresses on L4 APIs.
+
+## Goals
+
+- add matching rules for address to `TCPRoute`
+- add matching rules for address to `UDPRoute`
+- intentionally avoid type definitions that would make it hard to expand later
+
+## Non-Goals
+
+- define rules for port matching
+
+## Introduction
+
+While `TCPRoute` and `UDPRoute` currently support custom matching extensions,
+there is desire among the community to include some "fundamental" matching
+options in the spec that cover the most common requirements. In this GEP we
+request address matching for these APIs in order to support a standard
+for some of the commonplace setups of gateway implementations. Matching is
+intended to be covered for both _source_ and _destination_ to enable a finer
+level of tuning options for L4 traffic routing at a level below the `Gateway`.
+
+## API
+
+The API changes include the following new types:
+
+- `AddressMatch` to indicate the IP for address matching
+- `TrafficMatches` to configure matching according to network address
+
+These types enable the address matching required, with some active
+considerations about how to leave these open ended for later expansion.
+
+### AddressMatch Type
+
+A new `AddressMatch` type provides the targeting mechanism for match inclusion
+of a given network address:
+
+```go
+type AddressMatch struct {
+	// Type of the address, either IPAddress or NamedAddress.
+	//
+	// If NamedAddress is used this is a custom and specific value for each
+	// implementation to handle (and add validation for) according to their
+	// own needs.
+	//
+	// For IPAddress the implementor may expect either IPv4 or IPv6.
+	//
+	// +optional
+	// +kubebuilder:validation:Enum=IPAddress;NamedAddress
```

Comment:
Another option that would allow more experimentation by implementations is to have this Type field be a named string type, that we provide standard constants for, but that can also accept a domain-prefixed string. The constants we export would have a defined support policy, anything domain-prefixed is ImplementationSpecific.

So, we would make `IPAddress` available now for sure (with Extended or even Core support), but we can leave `NamedAddress` out if we want - implementations are free to build out their own `implementation.io/NamedAdddress` behavior, and then bring it back to us for Extended support consideration.

This avoids the version-incompatibilty problem of adding values to an enum being a breaking change, at the cost of a bit more complexity, and less standardization.

(This is cribbed from the metav1.Condition `Type` field, which has similar concerns).

Reply:
I feel generally good about this idea, once I have gathered some more feedback above about `NamedAddress` I'll address this further and ideally implement it.

Reply:
I do like the idea of vendor prefixed names here like `example.com/NamedAddress`, but I'm struggling to find additional use cases beyond NamedAddress. Since we already have NamedAddress elsewhere in the API (Gateway Spec+Status), it could get confusing to have different ways to represent that concept. If there were other types of addresses we might want to match on, I think domain prefixed types could still make sense though.

Reply:
As per our discussion in the sync today I created https://github.com/kubernetes-sigs/gateway-api/issues/958 as a follow up to do this holistically (e.g. for `Gateway` as well) so that we can use what exists for now to move forward, but make sure to reconsider how this works before `v1beta1`.